src/adapters/ethereum.ts

```diff
@@ -11,4 +16,149 @@ const getConnectedAccounts = (provider: Web3Provider): Promise<string[]> => {
+  provider: Web3Provider;
+  token: Token;
+}
+
+const isPermitSupported = async ({
+  provider,
+  token,
+}: IsPermitSupportedParams): Promise<boolean> => {
+  const tokenContractWithPermit = Erc20Permit__factory.connect(token.address, provider.getSigner());
+
+  try {
+    return !!(await tokenContractWithPermit.DOMAIN_SEPARATOR());
+  } catch (err) {
+    return false;
+  }
+};
+
+interface ApproveParams {
+  token: Token;
+  amount: BigNumber;
+  provider: Web3Provider;
+  owner: string;
+  spender: string;
+}
+
+const approve = async ({
+  token,
+  amount,
+  provider,
+  owner,
+  spender,
+}: ApproveParams): Promise<void> => {
+  if (token.address === ethersConstants.AddressZero) {
+    throw new Error("Cannot perfom an approve on ETH");
+  }
+
+  const erc20Contract = Erc20__factory.connect(token.address, provider.getSigner());
+  const hasAllowance = await isContractAllowedToSpendToken({
+    token,
+    amount,
+    provider,
+    owner,
+    spender,
+  });
+
+  if (!hasAllowance) {
+    const tx = await erc20Contract.approve(spender, ethersConstants.MaxUint256);
+
+    await tx.wait();
+  }
+};
+
+interface IsContractAllowedToSpendTokenParams {
+  token: Token;
+  amount: BigNumber;
+  provider: JsonRpcProvider;
+  owner: string;
+  spender: string;
+}
+
+const isContractAllowedToSpendToken = async ({
+  token,
+  amount,
+  provider,
+  owner,
+  spender,
+}: IsContractAllowedToSpendTokenParams): Promise<boolean> => {
+  if (token.address === ethersConstants.AddressZero) {
+    throw new Error("Cannot check the allowance of ETH");
+  }
+
+  const erc20Contract = Erc20__factory.connect(token.address, provider);
+  const allowance = await erc20Contract.allowance(owner, spender);
+
+  return allowance.gte(amount);
+};
+
+interface PermitParams {
+  token: Token;
+  provider: Web3Provider;
+  owner: string;
+  spender: string;
+  value: BigNumber;
+}
+
+const permit = async ({
+  token,
+  provider,
+  owner,
+  spender,
+  value,
+}: PermitParams): Promise<string> => {
+  if (token.address === ethersConstants.AddressZero) {
+    return Promise.reject(new Error("Cannot perfom an approve on ETH"));
+  }
+
+  const erc20PermitContract = Erc20Permit__factory.connect(token.address, provider.getSigner());
+  const nonce = await erc20PermitContract.nonces(owner);
+  const deadline = ethersConstants.MaxUint256;
+  const chainId = (await provider.getNetwork()).chainId;
+  const domain = {
+    name: token.name,
+    version: "1",
+    chainId: chainId,
+    verifyingContract: token.address,
+  };
+  const types = {
+    Permit: [
+      { name: "owner", type: "address" },
+      { name: "spender", type: "address" },
+      { name: "value", type: "uint256" },
+      { name: "nonce", type: "uint256" },
+      { name: "deadline", type: "uint256" },
+    ],
+  };
+  const values = {
+    owner,
+    spender,
+    value,
+    nonce,
+    deadline,
+  };
+
+  const signature = await provider.getSigner()._signTypedData(domain, types, values);
```

Comment:
Well, as we are using Typescript, we would detect this change easily when building, so I won't stick to the `5.6.9` version... In fact, it would be great to know ASAP when this function become stable to replace it.

Reply:
You're right about TS! I hadn't even thought about it.